### PR TITLE
[TAPE] Brings the new QNode in-line with PennyLane core

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

* The tape refactor, so far, has created an 'ideal' tape. This PR instead introduces some compatibility fixes to ensure that the new QNode has the same behaviour as the old QNode. 

**Description of the Change:**

* The Autograd interface now treats all objects without `requires_grad=True` as *non-differentiable by default*, matching PyTorch. This way, operations with scalar inputs such as `qml.RX(0.5, wires=0)` will be treated as non-differentiable by default.

* In addition to the above change, the new QNode will automatically wrap all QNode positional arguments with `requires_grad=True` in the Autograd interface, unless `requires_grad=False`.

  While this may seem odd in light of the previous bullet point, these two changes cause Autograd QNodes to match the core Autograd QNode in terms of functionality: All positional arguments are treated as differentiable by default, even python lists and scalars.

  At some point, when we are okay breaking the demos, we should remove this hotfix from the QNode, and have it match the autograd quantum tape.

* The `pennylane.numpy.tensor` class, a subclass of `np.ndarray`, now *always* returns a `tensor` object when acted upon by numpy functions that would previously return Python scalars:

  ```pycon
  >>> np.sum(np.array([0.1, 0.2], requires_grad=False))
  tensor(0.3, requires_grad=False)
  >>> np.array([0.1, 0.2], requires_grad=True)[0]
  tensor(0.1, requires_grad=True)
  ```

  This ensures that information about parameter differentiability is not lost upon functions that accept arrays and return scalars. This is a required fix for the previous two bullet points to work correctly.

* The autograd interface now unwraps all NumPy scalar arrays to Python literals before executing the device. This fixes an issue where plugin devices (such as `strawberryfields.fock`) that cannot accept scalar arrays (`np.array(0.5)`) as input would raise exceptions.

* Finally, QNodes that return output of size `[1]` will now always squeeze the output, returning a scalar. This matches the core QNode, which will always return a scalar for QNodes that return measurements of a single observable, e.g., `return expval(qml.PauliZ(0))`.

**Benefits:**

* The new QNode should now match the old QNode exactly in terms of behaviour.

**Possible Drawbacks:**

* We may decide to remove these compatibility hotfixes at some point, which will break existing tutorials and demos.

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/beta/interfaces/autograd.py
Comment (josh146) [inline pennylane/beta/interfaces/autograd.py:104]: On the forward pass, the user must specify `requires_grad=True` for parameters to be marked as default. Previously, arrays without this attribute, and all python numeric literals, would by default be treated as differentiable.

On the backward pass, we count the existence of an arraybox as differentiable.

@trbromley, this _might_ reduce the number of autograd computations?

File: pennylane/beta/interfaces/autograd.py
Comment (josh146) [inline pennylane/beta/interfaces/autograd.py:126]: @trbromley, this fixes the bug you noticed with the SF plugin (this bug did not affect built-in devices, hence was not picked up in the tests)

File: tests/beta/interfaces/test_qnode_autograd.py
Comment (josh146) [inline tests/beta/interfaces/test_qnode_autograd.py:51]: Since the QNode now returns a scalar by default

File: tests/beta/interfaces/test_qnode_autograd.py
Comment (josh146) [inline tests/beta/interfaces/test_qnode_autograd.py:72]: This example explains why

* the Autograd tape was changed to assume non-differentiable by default, and
* The QNode was changed to treat its positional arguments as differentiable by default.

This allows us to match the existing behaviour of the old QNode, where positional arguments were wrapped as `Variable` objects if `requires_grad != False`.

Here, the RX gate has a constant parameter _inside_ the QNode -- the tape treats this as non-differentiable. If we were to instead do

```python
@qnode(dev, interface="autograd")
def circuit(a, b):
    qml.RY(a, wires=0)
    qml.RX(b, wires=0)
    return expval(qml.PauliZ(0))
```
and we called `circuit(0.1, 0.2)`, then both rotation gates will now be trainable.

In effect, we are imitating the behaviour where the presence of a parameter in the signature makes that parameter differentiable, which will then propagate down to gate arguments.

**Note:** the above discussion applies _only_ to the autograd interface. The TF and Torch interfaces continue to work the same as always.

File: pennylane/beta/interfaces/autograd.py
Comment (mariaschuld) [reply pennylane/beta/interfaces/autograd.py:104]: Does that mean we now have to run a circuit forward before optimising it, like in pytorch? In other words, what makes the backward pass have the `ArrayBox`?

File: pennylane/beta/tapes/qnode.py
Comment (mariaschuld) [inline pennylane/beta/tapes/qnode.py:391]: I think it would be good to continue having "this parameter is trainable" as default, no?

Comment (mariaschuld) [inline pennylane/beta/tapes/qnode.py:391]: I think it would be quite helpful for future coders to rephrase "to maintain compatibility with core", since after this is merged, the code _will_ be core and this could be confusing? Maybe just say "hotfix to treat all inputs..."?

File: pennylane/beta/tapes/qnode.py
Comment (mariaschuld) [inline pennylane/beta/tapes/qnode.py:404]: Same here

File: pennylane/beta/tapes/qnode.py
Comment (mariaschuld) [inline pennylane/beta/tapes/qnode.py:408]: Is this idiot proof? For example when the return type is in a sub module? (But I am not sure what's going on here, so ignore if so)

File: pennylane/beta/tapes/qnode.py
Comment (mariaschuld) [inline pennylane/beta/tapes/qnode.py:417]: What is our roadmap to removing these hotfixes and changing the qnode behaviour?

File: pennylane/beta/interfaces/autograd.py
Comment (josh146) [reply pennylane/beta/interfaces/autograd.py:104]: > Does that mean we now have to run a circuit forward before optimising it

Nope, it works perfectly fine simply defining the qnode, creating the grad function, and calling the grad function straight away.

>  what makes the backward pass have the ArrayBox?

I'm not entirely sure, just something I've noticed with how autograd works.

* When you call an Autograd function, it always evaluates it numerically.

* When you call the `grad_fn` of an autograd function, it evaluates the function with `ArrayBox` objects, and then back-propagates to construct the final gradient. I _think_ the presence of `ArrayBox`'s is to help with gradient accumulation? But I'm not entirely sure.

File: tests/beta/interfaces/test_qnode_autograd.py
Comment (mariaschuld) [reply tests/beta/interfaces/test_qnode_autograd.py:51]: Wait, is the output behaviour of a qnode changed to current master, or not? Just trying to understand why the test has to change then?

File: pennylane/beta/tapes/qnode.py
Comment (josh146) [reply pennylane/beta/tapes/qnode.py:391]: This is something that came up in the Variable refactor ADR. @co9olguy preferred if the wrapped NumPy library matched PyTorch, and was not trainable by default.

In the end, we decided to keep it trainable as default, to allow demos such as the qubit rotation one to continue to work. However, we marked this as something we wanted to revisit.

Comment (josh146) [reply pennylane/beta/tapes/qnode.py:391]: Good point!

File: pennylane/beta/tapes/qnode.py
Comment (josh146) [reply pennylane/beta/tapes/qnode.py:408]: Here is what it looks like for the (known) datatypes we will have:

```python
In [48]: np.array([0.1]).__class__.__module__
Out[48]: 'numpy'

In [49]: torch.tensor([0.1]).__class__.__module__
Out[49]: 'torch'

In [50]: tf.constant([0.1]).__class__.__module__
Out[50]: 'tensorflow.python.framework.ops'

In [51]: qml.numpy.array([0.1]).__class__.__module__
Out[51]: 'pennylane.numpy.tensor'
```

The `.split(".")` drops all submodules

Comment (josh146) [reply pennylane/beta/tapes/qnode.py:408]: This is definitely a bit of a 'hack', but unfortunately we can't rely on interfaces, since the underlying device might be a 'backprop' device that natively returns a specific datatype

File: pennylane/numpy/tensor.py
Comment (mariaschuld) [inline pennylane/numpy/tensor.py:123]: Just to understand, why are these needed now? or is this just for convenience?

File: pennylane/beta/tapes/qnode.py
Comment (mariaschuld) [inline pennylane/beta/tapes/qnode.py:415]: Could this not be done in the interfaces, where they overwrite the execute function?

File: pennylane/beta/tapes/qnode.py
Comment (josh146) [reply pennylane/beta/tapes/qnode.py:417]: It depends on when we want to break autograd behaviour :) in the ADR, we punted this off as a 'future decision'.

File: pennylane/numpy/tensor.py
Comment (josh146) [reply pennylane/numpy/tensor.py:123]: This is required so that requires_grad info is not lost:
```python
>>> np.array([0.1, 0.2], requires_grad=True)[0]
np.array(0.1, requires_grad=True)
```

This is actually something we should have in core anyway, @co9olguy pointed this out while writing a demo recently.

File: pennylane/beta/tapes/qnode.py
Comment (mariaschuld) [reply pennylane/beta/tapes/qnode.py:404]: Do I understand correctly, the new qnode would find it more natural to have unsqueezed outputs, but the old qnode uses squeezed ones? Which option is better? Which one do we want to support in the long term?

File: pennylane/beta/tapes/qnode.py
Comment (mariaschuld) [reply pennylane/beta/tapes/qnode.py:391]: I assume you have thought about it, but this could not be done in the parameter update function defined by the autograd tape, right?

File: pennylane/beta/tapes/qnode.py
Comment (josh146) [reply pennylane/beta/tapes/qnode.py:415]: I originally thought the same, and had tried to do this.

I added `._squeeze` attributes to the interfaces, and had `AutogradInterface._squeeze(res)` here.

Unfortunately, this doesn't work for 'passthru' devices though, since they don't have an 'interface', they are purely written in the ML framework already!

File: tests/beta/interfaces/test_qnode_tf.py
Comment (mariaschuld) [inline tests/beta/interfaces/test_qnode_tf.py:53]: Is this comment still valid after the change?

File: tests/beta/interfaces/test_qnode_tf.py
Comment (mariaschuld) [inline tests/beta/interfaces/test_qnode_tf.py:59]: Is this because now every parameter in the tape is assumed to be non-trainable, unless otherwise stated (see above)?

**Review by mariaschuld — APPROVED** [review]
I approve this because my comments are all just questions, and I don't want to delay the refactor. However,  you may want to wait for a second reviewer, since I am not sure I understand the context of these hotfixes so well. 

I assume they just change the shape of the output and the assumptions about trainability, right?

File: pennylane/beta/tapes/qnode.py
Comment (josh146) [reply pennylane/beta/tapes/qnode.py:404]: Personally, I find it cleaner if the QNode always returned an array. A single measurement simply returns an array of length 1.

_However_ we have two many tutorials/demos that use the fact that a single measurement returns a scalar, so I am hesitant to make a change to this behaviour as part of the refactor. The refactor should ideally leave the UI and API unchanged, and just modify the internals.

But we should re-visit this and decide if we want to change the behaviour. We could do it in stages; make the squeezing optional, and being to show a deprecation warning.

File: pennylane/numpy/tensor.py
Comment (mariaschuld) [reply pennylane/numpy/tensor.py:123]: Ah great, I see!

File: pennylane/beta/tapes/qnode.py
Comment (mariaschuld) [reply pennylane/beta/tapes/qnode.py:415]: I always forget about those. Sure!

**Comment by josh146** [comment]
Thanks @mariaschuld! Your questions definitely helped.

> I assume they just change the shape of the output and the assumptions about trainability, right?

Yep. It simply changes the shape of the output for single measurements, and assumptions about (autograd) trainability.

File: pennylane/beta/tapes/qnode.py
Comment (mariaschuld) [reply pennylane/beta/tapes/qnode.py:415]: An alternative would be to give passthru devices a dummy interface that just does some post-processing...just a thought. So everything is on the same footing.

File: pennylane/numpy/tensor.py
Comment (josh146) [reply pennylane/numpy/tensor.py:123]: It also now matches TF and Torch:

image

**Comment by mariaschuld** [comment]
> Thanks @mariaschuld! Your questions definitely helped.
> 
> > I assume they just change the shape of the output and the assumptions about trainability, right?
> 
> Yep. It simply changes the shape of the output for single measurements, and assumptions about (autograd) trainability. This is a precursor PR to get the new QNode ready to run on the existing core tests.

Cool. I also only read your great introduction now - somehow escaped me. A lot of questions are answered there.

**Comment by mariaschuld** [comment]
Since this pr merges into a holding branch, feel free to merge - we can fix things later if needed, also using the tests!

File: pennylane/beta/tapes/qnode.py
Comment (josh146) [reply pennylane/beta/tapes/qnode.py:415]: Not sure if it is worth the extra overhead/maintenance... the main purpose of the interfaces is to provide custom gradient wrappers, which the backprop devices do not need

File: pennylane/beta/tapes/qnode.py
Comment (josh146) [reply pennylane/beta/tapes/qnode.py:391]: Yep. We actually want different behaviour between the autograd tape and the autograd QNode, to match their UI.

* If using an autograd tape, and we have a gate with a scalar parameter `0.6`, we _don't_ want to treat this as a differentiable parameter. We only want to consider `np.array(0.6, requires_grad=True)` as trainable.

* If we are using the QNode, and the user passes `0.6` as a positional QNode parameter, then we _do_ want to treat this as trainable, to match the existing core behaviour and to ensure all of our demos keep working. The hotfix here simply wraps all positional arguments with `requires_grad=True`, and pass this to the tape. (Note: positional parameters that explicitly have `requires_grad=False` will be skipped).

**Comment by josh146** [comment]
Update: rather than replacing core, the plan is now to have this sit alongside core. So I will work on getting all tests to pass before this is merged in.

File: pennylane/beta/devices/default_tensor.py
Comment (josh146) [inline pennylane/beta/devices/default_tensor.py:354]: Since indexing into a `qml.numpy` array now always returns an array, in several places we now need to explicitly cast to lists if we do _not_ want the output to remain an array.

File: pennylane/templates/subroutines/uccsd.py
Comment (josh146) [inline pennylane/templates/subroutines/uccsd.py:176]: It is now possible for the basis state to be of the form
```python
[np.array(1), np.array(0), ...]
```
so we need to allow for this.

File: pennylane/beta/interfaces/autograd.py
Comment (josh146) [reply pennylane/beta/interfaces/autograd.py:104]: `getattr(p, "requires_grad", True)` would previously have the logic:

* If `isinstance(p, qml.numpy.tensor)`, then it would simply use the specified version of `p.requires_grad`.

* If `not isinstance(p, qml.numpy.tensor)` (for example, if `p` was a float or a list, and so didn't have a `requires_grad` attribute), the tape would previously continue to assume it was a trainable parameter.

The change here only affects the second bullet point. With this change, non-PL numpy arrays, including floats, tuples, lists, etc., are now assumed to be constant by default. If you want to make a float trainable, you _have_ to now do `np.array(0.564, requires_grad=True)`.

Note that the above applies only to the (autograd) tape directly, the (autograd) QNode has different logic.

File: pennylane/beta/tapes/qnode.py
Comment (josh146) [reply pennylane/beta/tapes/qnode.py:403]: 🤯 I... did not think of this! My mind instantly went straight to 'we need to dispatch to `fn.squeeze()`.

Let me see if this works!

File: pennylane/wires.py
Comment (josh146) [reply pennylane/wires.py:37]: I suppose the key here, is that if `wires` does not have a `shape` attribute, we don't want this `elif` statement to evaluate as `True`. So the default must be anything that causes the equality to fail; it can be anything _except_ `tuple()`.

Originally, I had this more explicit:
```python
elif hasattr(wires, "shape") and wires.shape == tuple():
```
I reduced it to the above because I was like 'shorter is more elegant'. But, perhaps better to make it explicit!

Comment (josh146) [reply pennylane/wires.py:37]:
```suggestion
    elif hasattr(wires, "shape") and wires.shape == tuple():
```

**Comment by josh146** [comment]
@trbromley thank you for your `res[0]` suggestion, turns out I was massively overthinking it! All the tests pass, so I've removed the module introspection approach

File: pennylane/beta/interfaces/autograd.py
Comment (trbromley) [reply pennylane/beta/interfaces/autograd.py:104]: Ohhhhh! I was reading it as `getattr(p, "requires_grad") is False` :facepalm: Ok now I agree it looks good! Thanks.

File: pennylane/beta/tapes/qnode.py
Comment (trbromley) [inline pennylane/beta/tapes/qnode.py:423]: I wonder if this could have any performance implications (e.g. reporting a module on each call?). Is importing a heavy operation?

Comment (trbromley) [reply pennylane/beta/tapes/qnode.py:423]: I couldn't see any noticeable difference when testing the autograd and torch interfaces, so all good!

**Review by trbromley — APPROVED** [review]

File: pennylane/beta/tapes/qnode.py
Comment (josh146) [reply pennylane/beta/tapes/qnode.py:423]: > I wonder if this could have any performance implications (e.g. importing a module on each call?). Is importing a heavy operation?

I don't believe so. When Python imports a library, it remains in memory and is never re-imported on subsequent calls within the same session. It's why we can have `import tensorflow` in multiple PennyLane modules, and it never causes performance issues!

Comment (josh146) [reply pennylane/beta/tapes/qnode.py:423]: And in this case, if the tape is returning a autograd/torch/tf datatype, it means that autograd/torch/tf have _already_ been imported somewhere earlier in the pipeline :)

Comment (trbromley) [reply pennylane/beta/tapes/qnode.py:423]: Thanks, that's great! Explains why the performance times remained the same

**Pull request merged**
